java/arcs/core/host/AbstractArcHost.kt

```diff
@@ -10,43 +10,403 @@
+                if (particleState in alreadySucceededOnCreateStates) {
+                    particleState = ParticleState.Created
+                } else {
+                    particleState = ParticleState.Instantiated
+                }
+            }
+        }
+    }
+
+    /**
+     * Look up an existing [ParticleContext] in the current [ArcHostContext] if it exists for
+     * the specified [Plan.Particle] and [Particle], otherwise create and initialize a new
+     * [ParticleContext].
+     */
+    fun lookupParticleContextOrCreate(
+        context: ArcHostContext,
+        spec: Plan.Particle,
+        particle: Particle
+    ) = context.particles[spec]?.copy(particle = particle) ?: ParticleContext(particle)
+
+    /**
+     * Invokes any necessary lifecycle methods for the current [ArcHostContext.arcState] and any
+     * [Particle]s it may refer to, and may alter the current [ArcState].
+     */
+    suspend fun performLifecycleForContext(context: ArcHostContext) {
+        context.particles.values.forEach { particleContext ->
+            performParticleLifecycle(particleContext)
+            if (particleContext.particleState == ParticleState.Failed ||
+                particleContext.particleState == ParticleState.MaxFailed) {
+                context.arcState = ArcState.Error
+                return@forEach
+            }
+        }
+
+        if (context.arcState != ArcState.Error) {
+            context.arcState = ArcState.Running
+        }
+    }
+
+    /**
+     * Invokes necessary [Particle] lifecycle methods given the current
+     * [ParticleContext.particleState], and changes that state if necessary. For example by
+     * insuring that [Particle.onCreate()], [Particle.onShutdown()] are properly called.
+     */
+    suspend fun performParticleLifecycle(particleContext: ParticleContext) {
+        if (particleContext.particleState == ParticleState.Instantiated) {
+            try {
+                // onCreate() must succeed, else we consider the particle startup failed
+                particleContext.particle.onCreate()
+                particleContext.particleState = ParticleState.Created
+            } catch (e: Exception) {
+                markParticleAsFailed(particleContext)
+                return
+            }
+        }
+
+        // Should only happen if host crashes, restarts, and last persisted state was Running
+        if (particleContext.particleState == ParticleState.Started) {
+            particleContext.particleState == ParticleState.Stopped
+        }
+
+        // If we reach here, particle is being restarted
+        if (particleContext.particleState == ParticleState.Stopped) {
+            particleContext.particleState = ParticleState.Created
+        }
+
+        // This is temporary until the BaseParticle PR lands and onStartup() API lands.
+        // We force sync() calls in lieu of onStartup() API for demos
```

Comment:
Could you rephrase this as a TODO? It's not clear to me exactly what needs to change.